osu.Game/Overlays/SkinEditor/SkinEditorChangeHandler.cs

```diff
@@ -56,20 +57,51 @@ protected override void ApplyStateChange(byte[] previousState, byte[] newState)
             if (deserializedContent == null)
                 return;
 
-            SerialisedDrawableInfo[] skinnableInfo = deserializedContent.ToArray();
-            Drawable[] targetComponents = firstTarget.Components.OfType<Drawable>().ToArray();
+            SerialisedDrawableInfo[] skinnableInfos = deserializedContent.ToArray();
+            ISerialisableDrawable[] targetComponents = firstTarget.Components.ToArray();
 
-            if (!skinnableInfo.Select(s => s.Type).SequenceEqual(targetComponents.Select(d => d.GetType())))
+            // Store components based on type for later lookup
+            var typedComponents = new Dictionary<Type, List<ISerialisableDrawable>>();
+
+            foreach (var component in targetComponents)
             {
-                // Perform a naive full reload for now.
-                firstTarget.Reload(skinnableInfo);
+                Type lookup = component.GetType();
+
+                if (!typedComponents.TryGetValue(lookup, out List<ISerialisableDrawable>? typeComponents))
+                    typedComponents.Add(lookup, typeComponents = new List<ISerialisableDrawable>());
+
+                typeComponents.Add(component);
             }
-            else
+
+            // Remove all components
+            for (int i = targetComponents.Length - 1; i >= 0; i--)
+                firstTarget.Remove(targetComponents[i], false);
+
+            // Keeps count of how many components for each type were already revived
+            Dictionary<Type, int> typedComponentCounter = typedComponents.Keys.ToDictionary(t => t, _ => 0);
+
+            foreach (var skinnableInfo in skinnableInfos)
             {
-                int i = 0;
+                Type lookup = skinnableInfo.Type;
+
+                if (!typedComponents.TryGetValue(lookup, out List<ISerialisableDrawable>? typeComponents))
+                {
+                    firstTarget.Add((ISerialisableDrawable)skinnableInfo.CreateInstance());
+                    continue;
+                }
+
+                int typeComponentsUsed = typedComponentCounter[lookup]++;
+
+                ISerialisableDrawable component;
+
+                if (typeComponentsUsed < typeComponents.Count)
+                    // Re-use unused component
+                    ((Drawable)(component = typeComponents[typeComponentsUsed])).ApplySerialisedInfo(skinnableInfo);
+                else
+                    // Create new one
+                    component = (ISerialisableDrawable)skinnableInfo.CreateInstance();
```

Comment:
This doesn't need to be this complicated (the stuff with the two dictionaries and the counting). You have a mutable lookup (realised as a `Dictionary<Type, List<ISerialisableDrawable>>`, so use it instead rather than counting manually:

```csharp
if (typeComponents[lookup].Any())
{
    component = typeComponents.Remove(typeComponents.Count - 1);
    ((Drawable)component).ApplySerialisedInfo(skinnableInfo);
}
else
{
    component = (ISerialisableDrawable)skinnableInfo.CreateInstance();
}
```

Hell if you made it a `Dictionary<Type, Stack<ISerialisableDrawable>>` then you wouldn't need to faff about with the indices either, you would just `.Pop()` an instance for reuse.